File: src/m3tm/mobile/android/sdk/java/com/m3tm/sdk/security/SecureStorageManager.java
package com.m3tm.sdk.security;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyProperties;
import android.util.Log;

import androidx.annotation.RequiresApi;
import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKey;

import com.m3tm.sdk.M3TMException;

import java.io.IOException;
import java.security.GeneralSecurityException;
import java.security.KeyStore;
import java.security.SecureRandom;
import java.util.Arrays;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;

/**
 * OWASP MASVS-STORAGE uyumlu güvenli depolama yöneticisi.
 * 
 * Android KeyStore ve EncryptedSharedPreferences kullanarak
 * hassas verilerin güvenli şekilde saklanmasını sağlar.
 */
public class SecureStorageManager {
    private static final String TAG = "SecureStorageManager";
    private static final String PREFS_NAME = "m3tm_secure_prefs";
    private static final String KEYSTORE_ALIAS = "M3TMSecureStorageKey";
    private static final String ANDROID_KEYSTORE = "AndroidKeyStore";
    private static final int GCM_IV_LENGTH = 12;
    private static final int GCM_TAG_LENGTH = 16;
    
    private final Context context;
    private SharedPreferences encryptedPrefs;
    private MasterKey masterKey;
    private boolean isInitialized = false;
    private boolean hardwareBackedSecurity = false;
    
    public SecureStorageManager(Context context) {
        this.context = context.getApplicationContext();
    }
    
    /**
     * Güvenli depolama sistemini başlatır.
     * Hardware-backed keystore kullanmayı dener, başarısız olursa software fallback kullanır.
     * 
     * @throws M3TMException Başlatma başarısız olursa
     */
    public void initialize() throws M3TMException {
        try {
            // Hardware-backed keystore kontrolü
            hardwareBackedSecurity = checkHardwareBackedSecurity();
            Log.i(TAG, "Hardware-backed security: " + hardwareBackedSecurity);
            
            // MasterKey oluştur (hardware-backed preferred)
            masterKey = createMasterKey();
            
            // EncryptedSharedPreferences oluştur
            encryptedPrefs = EncryptedSharedPreferences.create(
                context,
                PREFS_NAME,
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            );
            
            isInitialized = true;
            Log.i(TAG, "Secure storage initialized successfully");
            
        } catch (GeneralSecurityException | IOException e) {
            Log.e(TAG, "Failed to initialize secure storage", e);
            throw new M3TMException(M3TMException.ErrorCode.INITIALIZATION_ERROR, 
                "Secure storage initialization failed", e);
        }
    }
    
    /**
     * Hardware-backed security desteğini kontrol eder.
     * 
     * @return true eğer hardware-backed keystore mevcutsa
     */
    private boolean checkHardwareBackedSecurity() {
        try {
            KeyStore keyStore = KeyStore.getInstance(ANDROID_KEYSTORE);
            keyStore.load(null);
            
            // Test key oluşturarak hardware-backed support kontrolü
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                KeyGenerator keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE);
                KeyGenParameterSpec.Builder builder = new KeyGenParameterSpec.Builder(
                    "test_hw_key",
                    KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
                    .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE);
                
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    builder.setIsStrongBoxBacked(true); // Hardware security module
                }
                
                keyGenerator.init(builder.build());
                SecretKey testKey = keyGenerator.generateKey();
                
                // Test key'i sil
                keyStore.deleteEntry("test_hw_key");
                
                return testKey != null;
            }
            return false;
        } catch (Exception e) {
            Log.w(TAG, "Hardware-backed security check failed", e);
            return false;
        }
    }
    
    /**
     * MasterKey oluşturur, hardware-backed keystore tercih eder.
     * 
     * @return MasterKey instance
     * @throws GeneralSecurityException
     */
    private MasterKey createMasterKey() throws GeneralSecurityException {
        MasterKey.Builder builder = new MasterKey.Builder(context, KEYSTORE_ALIAS)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM);
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P && hardwareBackedSecurity) {
            builder.setRequestStrongBoxBacked(true);
        }
        
        try {
            return builder.build();
        } catch (Exception e) {
            Log.w(TAG, "Failed to create hardware-backed master key, falling back to software", e);
            // Software fallback
            return new MasterKey.Builder(context, KEYSTORE_ALIAS + "_sw")
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .setRequestStrongBoxBacked(false)
                .build();
        }
    }
    
    /**
     * Güvenli string depolama.
     * 
     * @param key Anahtar
     * @param value Değer
     * @throws M3TMException Depolama başarısız olursa
     */
    public void storeSecureString(String key, String value) throws M3TMException {
        checkInitialized();
        
        try {
            encryptedPrefs.edit()
                .putString(key, value)
                .apply();
            
            Log.d(TAG, "Secure string stored for key: " + key);
            
        } catch (Exception e) {
            Log.e(TAG, "Failed to store secure string", e);
            throw new M3TMException(M3TMException.ErrorCode.STORAGE_ERROR, 
                "Failed to store secure data", e);
        }
    }
    
    /**
     * Güvenli string okuma.
     * 
     * @param key Anahtar
     * @param defaultValue Varsayılan değer
     * @return Saklanan değer veya varsayılan değer
     * @throws M3TMException Okuma başarısız olursa
     */
    public String getSecureString(String key, String defaultValue) throws M3TMException {
        checkInitialized();
        
        try {
            return encryptedPrefs.getString(key, defaultValue);
            
        } catch (Exception e) {
            Log.e(TAG, "Failed to get secure string", e);
            throw new M3TMException(M3TMException.ErrorCode.STORAGE_ERROR, 
                "Failed to retrieve secure data", e);
        }
    }
    
    /**
     * Güvenli byte array depolama.
     * AES-GCM-256 ile şifrelenerek saklanır.
     * 
     * @param key Anahtar
     * @param data Şifrelenecek veri
     * @throws M3TMException Şifreleme veya depolama başarısız olursa
     */
    public void storeSecureData(String key, byte[] data) throws M3TMException {
        checkInitialized();
        
        try {
            byte[] encryptedData = encryptWithAESGCM(data);
            String base64Data = android.util.Base64.encodeToString(encryptedData, android.util.Base64.DEFAULT);
            
            encryptedPrefs.edit()
                .putString(key + "_encrypted", base64Data)
                .apply();
            
            Log.d(TAG, "Secure data stored for key: " + key);
            
        } catch (Exception e) {
            Log.e(TAG, "Failed to store secure data", e);
            throw new M3TMException(M3TMException.ErrorCode.STORAGE_ERROR, 
                "Failed to store encrypted data", e);
        }
    }
    
    /**
     * Güvenli byte array okuma.
     * AES-GCM-256 ile çözümlenir.
     * 
     * @param key Anahtar
     * @return Çözümlenmiş veri, null eğer bulunamazsa
     * @throws M3TMException Çözümleme başarısız olursa
     */
    public byte[] getSecureData(String key) throws M3TMException {
        checkInitialized();
        
        try {
            String base64Data = encryptedPrefs.getString(key + "_encrypted", null);
            if (base64Data == null) {
                return null;
            }
            
            byte[] encryptedData = android.util.Base64.decode(base64Data, android.util.Base64.DEFAULT);
            return decryptWithAESGCM(encryptedData);
            
        } catch (Exception e) {
            Log.e(TAG, "Failed to get secure data", e);
            throw new M3TMException(M3TMException.ErrorCode.STORAGE_ERROR, 
                "Failed to retrieve encrypted data", e);
        }
    }
    
    /**
     * AES-GCM-256 ile veri şifreleme.
     * 
     * @param data Şifrelenecek veri
     * @return IV + şifrelenmiş veri + tag
     * @throws Exception Şifreleme başarısız olursa
     */
    private byte[] encryptWithAESGCM(byte[] data) throws Exception {
        // Secure random IV oluştur
        byte[] iv = new byte[GCM_IV_LENGTH];
        new SecureRandom().nextBytes(iv);
        
        // Cipher oluştur ve başlat
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(GCM_TAG_LENGTH * 8, iv);
        
        // KeyStore'dan key al
        KeyStore keyStore = KeyStore.getInstance(ANDROID_KEYSTORE);
        keyStore.load(null);
        SecretKey secretKey = (SecretKey) keyStore.getKey(KEYSTORE_ALIAS, null);
        
        cipher.init(Cipher.ENCRYPT_MODE, secretKey, gcmParameterSpec);
        
        // Şifrele
        byte[] encryptedData = cipher.doFinal(data);
        
        // IV + encrypted data + tag birleştir
        byte[] result = new byte[iv.length + encryptedData.length];
        System.arraycopy(iv, 0, result, 0, iv.length);
        System.arraycopy(encryptedData, 0, result, iv.length, encryptedData.length);
        
        return result;
    }
    
    /**
     * AES-GCM-256 ile veri çözümleme.
     * 
     * @param encryptedData IV + şifrelenmiş veri + tag
     * @return Çözümlenmiş veri
     * @throws Exception Çözümleme başarısız olursa
     */
    private byte[] decryptWithAESGCM(byte[] encryptedData) throws Exception {
        // IV'yi ayır
        byte[] iv = Arrays.copyOfRange(encryptedData, 0, GCM_IV_LENGTH);
        byte[] cipherText = Arrays.copyOfRange(encryptedData, GCM_IV_LENGTH, encryptedData.length);
        
        // Cipher oluştur ve başlat
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(GCM_TAG_LENGTH * 8, iv);
        
        // KeyStore'dan key al
        KeyStore keyStore = KeyStore.getInstance(ANDROID_KEYSTORE);
        keyStore.load(null);
        SecretKey secretKey = (SecretKey) keyStore.getKey(KEYSTORE_ALIAS, null);
        
        cipher.init(Cipher.DECRYPT_MODE, secretKey, gcmParameterSpec);
        
        // Çözümle
        return cipher.doFinal(cipherText);
    }
    
    /**
     * Anahtar silme.
     * 
     * @param key Silinecek anahtar
     */
    public void removeSecureData(String key) {
        if (!isInitialized) {
            return;
        }
        
        encryptedPrefs.edit()
            .remove(key)
            .remove(key + "_encrypted")
            .apply();
        
        Log.d(TAG, "Secure data removed for key: " + key);
    }
    
    /**
     * Tüm güvenli verileri temizle.
     */
    public void clearAllSecureData() {
        if (!isInitialized) {
            return;
        }
        
        encryptedPrefs.edit().clear().apply();
        Log.i(TAG, "All secure data cleared");
    }
    
    /**
     * Hardware-backed security durumunu döndürür.
     * 
     * @return true eğer hardware-backed keystore kullanılıyorsa
     */
    public boolean isHardwareBackedSecurity() {
        return hardwareBackedSecurity;
    }
    
    /**
     * Başlatılma durumunu kontrol eder.
     * 
     * @throws M3TMException Başlatılmamışsa
     */
    private void checkInitialized() throws M3TMException {
        if (!isInitialized) {
            throw new M3TMException(M3TMException.ErrorCode.NOT_INITIALIZED, 
                "SecureStorageManager not initialized");
        }
    }
    
    /**
     * Kaynakları serbest bırakır.
     */
    public void shutdown() {
        isInitialized = false;
        encryptedPrefs = null;
        masterKey = null;
        Log.i(TAG, "SecureStorageManager shutdown completed");
    }
}

File: src/m3tm/mobile/ios/sdk/SecureStorageManager.swift
import Foundation
import Security
import CommonCrypto

/**
 * OWASP MASVS-STORAGE uyumlu güvenli depolama yöneticisi (iOS).
 * 
 * iOS Keychain Services ve AES-GCM-256 şifreleme kullanarak
 * hassas verilerin güvenli şekilde saklanmasını sağlar.
 */
@objc public class SecureStorageManager: NSObject {
    
    // MARK: - Constants
    
    private static let serviceIdentifier = "com.m3tm.sdk.secure.storage"
    private static let encryptionKeyAlias = "M3TMSecureStorageKey"
    private static let gcmTagLength = 16
    private static let gcmIVLength = 12
    
    // MARK: - Properties
    
    private var isInitialized = false
    private var secureEnclaveAvailable = false
    private var encryptionKey: SecKey?
    
    // MARK: - Public Methods
    
    /**
     * Güvenli depolama sistemini başlatır.
     * Secure Enclave kullanmayı dener, başarısız olursa keychain fallback kullanır.
     * 
     * @param completion Başlatma sonucu callback'i
     */
    @objc public func initialize(completion: @escaping (Error?) -> Void) {
        DispatchQueue.global(qos: .utility).async { [weak self] in
            guard let self = self else {
                DispatchQueue.main.async {
                    completion(M3TMError.internalError("SecureStorageManager deallocated"))
                }
                return
            }
            
            do {
                // Secure Enclave availability check
                self.secureEnclaveAvailable = self.checkSecureEnclaveAvailability()
                NSLog("Secure Enclave available: \(self.secureEnclaveAvailable)")
                
                // Encryption key oluştur veya al
                try self.setupEncryptionKey()
                
                self.isInitialized = true
                NSLog("Secure storage initialized successfully")
                
                DispatchQueue.main.async {
                    completion(nil)
                }
                
            } catch {
                NSLog("Failed to initialize secure storage: \(error)")
                DispatchQueue.main.async {
                    completion(M3TMError.internalError("Secure storage initialization failed: \(error.localizedDescription)"))
                }
            }
        }
    }
    
    /**
     * Secure Enclave desteğini kontrol eder.
     * 
     * @return true eğer Secure Enclave mevcutsa
     */
    private func checkSecureEnclaveAvailability() -> Bool {
        if #available(iOS 9.0, *) {
            // Test key oluşturarak Secure Enclave support kontrolü
            let attributes: [String: Any] = [
                kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
                kSecAttrKeySizeInBits as String: 256,
                kSecAttrTokenID as String: kSecAttrTokenIDSecureEnclave,
                kSecPrivateKeyAttrs as String: [
                    kSecAttrAccessControl as String: SecAccessControlCreateWithFlags(
                        kCFAllocatorDefault,
                        kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
                        .privateKeyUsage,
                        nil
                    )!
                ]
            ]
            
            var error: Unmanaged<CFError>?
            let privateKey = SecKeyCreateRandomKey(attributes as CFDictionary, &error)
            
            if let key = privateKey {
                // Test key'i sil
                let deleteQuery: [String: Any] = [
                    kSecClass as String: kSecClassKey,
                    kSecAttrApplicationTag as String: "test_se_key".data(using: .utf8)!
                ]
                SecItemDelete(deleteQuery as CFDictionary)
                return true
            }
        }
        return false
    }
    
    /**
     * Encryption key'i oluşturur veya mevcut olanı alır.
     * Secure Enclave tercih eder, fallback olarak keychain kullanır.
     */
    private func setupEncryptionKey() throws {
        // Önce mevcut key'i kontrol et
        if let existingKey = try getExistingEncryptionKey() {
            self.encryptionKey = existingKey
            return
        }
        
        // Yeni key oluştur
        let keyData = try generateRandomKey()
        try storeEncryptionKeyInKeychain(keyData)
        
        // Key'i memory'de tut (SecKey olarak değil, güvenlik için)
        // Gerçek implementasyonda key'i her kullanımda keychain'den alınabilir
    }
    
    /**
     * Mevcut encryption key'i keychain'den alır.
     * 
     * @return SecKey varsa, nil yoksa
     */
    private func getExistingEncryptionKey() throws -> SecKey? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.serviceIdentifier,
            kSecAttrAccount as String: Self.encryptionKeyAlias,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        if status == errSecSuccess, let keyData = result as? Data {
            // Key data'yı SecKey'e çevir (AES için)
            // Burada basitleştirme için keyData'yı direkt kullanacağız
            return nil // Gerçek implementasyonda SecKey döndürülür
        } else if status == errSecItemNotFound {
            return nil
        } else {
            throw M3TMError.bridgeError("Failed to retrieve encryption key from keychain")
        }
    }
    
    /**
     * Güvenli random key oluşturur.
     * 
     * @return 256-bit random key
     */
    private func generateRandomKey() throws -> Data {
        var keyData = Data(count: 32) // 256 bit
        let result = keyData.withUnsafeMutableBytes { mutableBytes in
            SecRandomCopyBytes(kSecRandomDefault, 32, mutableBytes.bindMemory(to: UInt8.self).baseAddress!)
        }
        
        if result != errSecSuccess {
            throw M3TMError.bridgeError("Failed to generate random key")
        }
        
        return keyData
    }
    
    /**
     * Encryption key'i keychain'e güvenli şekilde saklar.
     * 
     * @param keyData Saklanacak key data
     */
    private func storeEncryptionKeyInKeychain(_ keyData: Data) throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.serviceIdentifier,
            kSecAttrAccount as String: Self.encryptionKeyAlias,
            kSecValueData as String: keyData,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]
        
        // Secure Enclave varsa access control ekle
        if secureEnclaveAvailable {
            if #available(iOS 9.0, *) {
                let accessControl = SecAccessControlCreateWithFlags(
                    kCFAllocatorDefault,
                    kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
                    .biometryAny,
                    nil
                )
                if let ac = accessControl {
                    var secureQuery = query
                    secureQuery[kSecAttrAccessControl as String] = ac
                    let status = SecItemAdd(secureQuery as CFDictionary, nil)
                    if status != errSecSuccess {
                        throw M3TMError.bridgeError("Failed to store encryption key in keychain with Secure Enclave")
                    }
                    return
                }
            }
        }
        
        // Fallback: Normal keychain storage
        let status = SecItemAdd(query as CFDictionary, nil)
        if status != errSecSuccess {
            throw M3TMError.bridgeError("Failed to store encryption key in keychain")
        }
    }
    
    /**
     * Güvenli string depolama.
     * 
     * @param key Anahtar
     * @param value Değer
     * @param completion İşlem sonucu callback'i
     */
    @objc public func storeSecureString(_ key: String, value: String, completion: @escaping (Error?) -> Void) {
        guard isInitialized else {
            completion(M3TMError.notInitialized)
            return
        }
        
        DispatchQueue.global(qos: .utility).async { [weak self] in
            do {
                let valueData = value.data(using: .utf8) ?? Data()
                try self?.storeInKeychain(key: key, data: valueData)
                DispatchQueue.main.async {
                    completion(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    completion(error)
                }
            }
        }
    }
    
    /**
     * Güvenli string okuma.
     * 
     * @param key Anahtar
     * @param completion Sonuç callback'i (value, error)
     */
    @objc public func getSecureString(_ key: String, completion: @escaping (String?, Error?) -> Void) {
        guard isInitialized else {
            completion(nil, M3TMError.notInitialized)
            return
        }
        
        DispatchQueue.global(qos: .utility).async { [weak self] in
            do {
                if let data = try self?.getFromKeychain(key: key),
                   let value = String(data: data, encoding: .utf8) {
                    DispatchQueue.main.async {
                        completion(value, nil)
                    }
                } else {
                    DispatchQueue.main.async {
                        completion(nil, nil)
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    completion(nil, error)
                }
            }
        }
    }
    
    /**
     * Güvenli data depolama.
     * AES-GCM-256 ile şifrelenerek saklanır.
     * 
     * @param key Anahtar
     * @param data Şifrelenecek veri
     * @param completion İşlem sonucu callback'i
     */
    @objc public func storeSecureData(_ key: String, data: Data, completion: @escaping (Error?) -> Void) {
        guard isInitialized else {
            completion(M3TMError.notInitialized)
            return
        }
        
        DispatchQueue.global(qos: .utility).async { [weak self] in
            do {
                let encryptedData = try self?.encryptWithAESGCM(data) ?? Data()
                try self?.storeInKeychain(key: key + "_encrypted", data: encryptedData)
                DispatchQueue.main.async {
                    completion(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    completion(error)
                }
            }
        }
    }
    
    /**
     * Güvenli data okuma.
     * AES-GCM-256 ile çözümlenir.
     * 
     * @param key Anahtar
     * @param completion Sonuç callback'i (data, error)
     */
    @objc public func getSecureData(_ key: String, completion: @escaping (Data?, Error?) -> Void) {
        guard isInitialized else {
            completion(nil, M3TMError.notInitialized)
            return
        }
        
        DispatchQueue.global(qos: .utility).async { [weak self] in
            do {
                if let encryptedData = try self?.getFromKeychain(key: key + "_encrypted") {
                    let decryptedData = try self?.decryptWithAESGCM(encryptedData) ?? Data()
                    DispatchQueue.main.async {
                        completion(decryptedData, nil)
                    }
                } else {
                    DispatchQueue.main.async {
                        completion(nil, nil)
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    completion(nil, error)
                }
            }
        }
    }
    
    /**
     * Keychain'e veri saklama.
     * 
     * @param key Anahtar
     * @param data Veri
     */
    private func storeInKeychain(key: String, data: Data) throws {
        // Önce mevcut item'ı sil
        removeFromKeychain(key: key)
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.serviceIdentifier,
            kSecAttrAccount as String: key,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]
        
        let status = SecItemAdd(query as CFDictionary, nil)
        if status != errSecSuccess {
            throw M3TMError.bridgeError("Failed to store data in keychain for key: \(key)")
        }
    }
    
    /**
     * Keychain'den veri okuma.
     * 
     * @param key Anahtar
     * @return Veri varsa, nil yoksa
     */
    private func getFromKeychain(key: String) throws -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.serviceIdentifier,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        if status == errSecSuccess {
            return result as? Data
        } else if status == errSecItemNotFound {
            return nil
        } else {
            throw M3TMError.bridgeError("Failed to retrieve data from keychain for key: \(key)")
        }
    }
    
    /**
     * Keychain'den veri silme.
     * 
     * @param key Anahtar
     */
    private func removeFromKeychain(key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.serviceIdentifier,
            kSecAttrAccount as String: key
        ]
        
        SecItemDelete(query as CFDictionary)
    }
    
    /**
     * AES-GCM-256 ile veri şifreleme.
     * 
     * @param data Şifrelenecek veri
     * @return IV + şifrelenmiş veri + tag
     */
    private func encryptWithAESGCM(_ data: Data) throws -> Data {
        // Random IV oluştur
        var iv = Data(count: Self.gcmIVLength)
        let result = iv.withUnsafeMutableBytes { mutableBytes in
            SecRandomCopyBytes(kSecRandomDefault, Self.gcmIVLength, mutableBytes.bindMemory(to: UInt8.self).baseAddress!)
        }
        
        if result != errSecSuccess {
            throw M3TMError.bridgeError("Failed to generate random IV")
        }
        
        // Encryption key al (basitleştirme için sabit key kullanıyoruz)
        guard let keyData = try getEncryptionKeyData() else {
            throw M3TMError.bridgeError("Encryption key not available")
        }
        
        // AES-GCM encryption (CommonCrypto ile)
        var encryptedData = Data(count: data.count + Self.gcmTagLength)
        var encryptedLength = 0
        
        let status = keyData.withUnsafeBytes { keyBytes in
            iv.withUnsafeBytes { ivBytes in
                data.withUnsafeBytes { dataBytes in
                    encryptedData.withUnsafeMutableBytes { encryptedBytes in
                        CCCryptorGCM(
                            CCOperation(kCCEncrypt),
                            CCAlgorithm(kCCAlgorithmAES),
                            keyBytes.bindMemory(to: UInt8.self).baseAddress,
                            keyData.count,
                            ivBytes.bindMemory(to: UInt8.self).baseAddress,
                            iv.count,
                            nil, 0, // Additional authenticated data
                            dataBytes.bindMemory(to: UInt8.self).baseAddress,
                            data.count,
                            encryptedBytes.bindMemory(to: UInt8.self).baseAddress,
                            &encryptedLength
                        )
                    }
                }
            }
        }
        
        if status != kCCSuccess {
            throw M3TMError.bridgeError("AES-GCM encryption failed")
        }
        
        // IV + encrypted data + tag birleştir
        var result = Data()
        result.append(iv)
        result.append(encryptedData.prefix(encryptedLength))
        
        return result
    }
    
    /**
     * AES-GCM-256 ile veri çözümleme.
     * 
     * @param encryptedData IV + şifrelenmiş veri + tag
     * @return Çözümlenmiş veri
     */
    private func decryptWithAESGCM(_ encryptedData: Data) throws -> Data {
        guard encryptedData.count > Self.gcmIVLength + Self.gcmTagLength else {
            throw M3TMError.bridgeError("Invalid encrypted data length")
        }
        
        // IV ve cipher text'i ayır
        let iv = encryptedData.prefix(Self.gcmIVLength)
        let cipherText = encryptedData.suffix(from: Self.gcmIVLength)
        
        // Encryption key al
        guard let keyData = try getEncryptionKeyData() else {
            throw M3TMError.bridgeError("Encryption key not available")
        }
        
        // AES-GCM decryption
        var decryptedData = Data(count: cipherText.count)
        var decryptedLength = 0
        
        let status = keyData.withUnsafeBytes { keyBytes in
            iv.withUnsafeBytes { ivBytes in
                cipherText.withUnsafeBytes { cipherBytes in
                    decryptedData.withUnsafeMutableBytes { decryptedBytes in
                        CCCryptorGCM(
                            CCOperation(kCCDecrypt),
                            CCAlgorithm(kCCAlgorithmAES),
                            keyBytes.bindMemory(to: UInt8.self).baseAddress,
                            keyData.count,
                            ivBytes.bindMemory(to: UInt8.self).baseAddress,
                            iv.count,
                            nil, 0, // Additional authenticated data
                            cipherBytes.bindMemory(to: UInt8.self).baseAddress,
                            cipherText.count,
                            decryptedBytes.bindMemory(to: UInt8.self).baseAddress,
                            &decryptedLength
                        )
                    }
                }
            }
        }
        
        if status != kCCSuccess {
            throw M3TMError.bridgeError("AES-GCM decryption failed")
        }
        
        return decryptedData.prefix(decryptedLength)
    }
    
    /**
     * Encryption key data'yı keychain'den alır.
     * 
     * @return Key data varsa, nil yoksa
     */
    private func getEncryptionKeyData() throws -> Data? {
        return try getFromKeychain(key: Self.encryptionKeyAlias)
    }
    
    /**
     * Anahtar silme.
     * 
     * @param key Silinecek anahtar
     * @param completion İşlem sonucu callback'i
     */
    @objc public func removeSecureData(_ key: String, completion: @escaping (Error?) -> Void) {
        guard isInitialized else {
            completion(M3TMError.notInitialized)
            return
        }
        
        DispatchQueue.global(qos: .utility).async { [weak self] in
            self?.removeFromKeychain(key: key)
            self?.removeFromKeychain(key: key + "_encrypted")
            DispatchQueue.main.async {
                completion(nil)
            }
        }
    }
    
    /**
     * Tüm güvenli verileri temizle.
     * 
     * @param completion İşlem sonucu callback'i
     */
    @objc public func clearAllSecureData(completion: @escaping (Error?) -> Void) {
        guard isInitialized else {
            completion(M3TMError.notInitialized)
            return
        }
        
        DispatchQueue.global(qos: .utility).async { [weak self] in
            let query: [String: Any] = [
                kSecClass as String: kSecClassGenericPassword,
                kSecAttrService as String: Self.serviceIdentifier
            ]
            
            SecItemDelete(query as CFDictionary)
            
            DispatchQueue.main.async {
                completion(nil)
            }
        }
    }
    
    /**
     * Secure Enclave kullanılabilirlik durumunu döndürür.
     * 
     * @return true eğer Secure Enclave kullanılabilirse
     */
    @objc public var isSecureEnclaveAvailable: Bool {
        return secureEnclaveAvailable
    }
    
    /**
     * Kaynakları serbest bırakır.
     */
    @objc public func shutdown() {
        isInitialized = false
        encryptionKey = nil
        NSLog("SecureStorageManager shutdown completed")
    }
}
